### PR TITLE
Habit Tracker Part 2

### PR DESCRIPTION
**Pull request by ealbuerne**
@rjulianialtimetrik  Please check the updates made to the Habit Tracker project, this is Part 2 update.

### CODE REVIEW

File: src/aura/DailyContainerComponent/DailyContainerComponent.cmp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<aura:component >
+    <!-- ATTRIBUTES -->
+    <aura:attribute name="dailyTasks" type="Task__c[]"/>
+    <aura:attribute name="totalDailyTask" type="Integer"/>
```
Comment (rjulianialtimetrik) [inline src/aura/DailyContainerComponent/DailyContainerComponent.cmp]: About this component and some others that are very similar: 
* You should not have a container component for each type of task you have. You can define a "TaskColumn" component or something like that and simply have an attribute called "tasks" receiving the list of tasks. Then you just use another component to render the items (you already do this, but you will need some code to decide which component will be shown based on the record type).
* totalDailyTask (and similar attributes) should be made general (totalTasks).

File: src/aura/HabitTrackerComponent/HabitTrackerComponent.cmp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+<!--HANDLERS-->
+<aura:handler name="init" value="{!this}" action="{!c.doInit}"/>
+<aura:handler name="updateTask" event="c:updateTaskEvent"
+        action="{!c.handleTaskEvent}"/>
+
+<div class="slds-grid slds-gutters"> 
+
+    <div class="slds-col">  
+        <c:ToDoContainerComponent toDoTasks="{!v.toDoTasks}" totalToDoTask="{!v.toDoTasks.length}"/>
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponent.cmp]: You should have a single type of Column-like component as mentioned in a previous comment.

File: src/aura/HabitTrackerComponent/HabitTrackerComponentController.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+({
+	doInit: function (component, event, helper) {
+		helper.getTaskList(component);
+	},
+
+	handleTaskEvent: function (component, event, helper) {
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentController.js]: this method's name has no meaning. you should name the method giving some sort of hint about its purpose.

File: src/aura/HabitTrackerComponent/HabitTrackerComponentController.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+		var task = event.getParam("task");
+
+		if (task.Record_Type_Name__c === "Habit") {
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentController.js]: This is wrong in the concept. You should not re-use the same element for completely different purposes. If you are deleting a Task (whatever its type, doesn't matter if it's an habit) you should have an event for that; same goes for the need to update anything... if you are handling that with an event you have to define a specific one for that action.

File: src/aura/HabitTrackerComponent/HabitTrackerComponentController.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+			helper.deleteHabitTask(component, task);
+		} else {
+			helper.updateDailyToDoTask(component, task);
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentController.js]: also, I would suggest you create more generic methods. updateTask and deleteTask should be good and help you achieve your goal while preventing needless checks

File: src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+				console.log("Failed with state: " + state);
+			}
+		});
+
+		// Send action off to be executed
+		$A.enqueueAction(action);
+	},
+
+	deleteHabitTask: function (component, task) {
+		var action = component.get("c.deleteHabitTask");
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js]: same as mentioned before, you should have a generic "deleteTask" method that performs the action regardless of the task's type.

File: src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		action.setParams({
+			"task": task
+		});
+
+		action.setCallback(this, function (response) {
+			var state = response.getState();
+			if (state === "SUCCESS") {
+				component.set("v.habitTasks", response.getReturnValue());
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js]: you should not query from the DB when you are performing a Delete. You just need to find the item you are deleting inside the local collection in javascript and remove the element from the list/array.

File: src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+		var action = component.get("c.updateStatus");
+		action.setParams({
+			"task": task
+		});
+
+		action.setCallback(this, function (response) {
+			var state = response.getState();
+			if (state === "SUCCESS") {
+				if (task.Record_Type_Name__c === "To-Do") {
+					component.set("v.toDoTasks", response.getReturnValue());
```
Comment (rjulianialtimetrik) [inline src/aura/HabitTrackerComponent/HabitTrackerComponentHelper.js]: you should not query from the DB when you are performing an Update unless your code needs values that are automatically filled in the Database (like, for example,formula fields). You just need to make sure the record is updated in the local list (after a component.get call) and do a component.set of the same list.

File: src/aura/TaskItemComponent/TaskItemComponent.cmp [hunk 9]
```diff
@@ -0,0 +1,39 @@
+<aura:component controller="TaskController">
+	<!--ATTRIBUTES-->
+    <aura:attribute name="task" type="Task__c"/>
+    <aura:attribute name="isHabit" type="Boolean"/>
```
Comment (rjulianialtimetrik) [inline src/aura/TaskItemComponent/TaskItemComponent.cmp]: why are you adding "isHabit" as an attribute? you should be able to retrieve that information from the Task__C record itself

File: src/aura/TaskItemComponent/TaskItemComponentController.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+({
+
+	doInit: function (component, event, helper) {
+		helper.getPickListStatusValues(component);
+	},
+
+	clickDeleteHabitTask: function (component, event, helper) {
+		var task = component.get("v.task");
```
Comment (rjulianialtimetrik) [inline src/aura/TaskItemComponent/TaskItemComponentController.js]: again, please have one event for delete actions and one for update.

File: src/classes/TaskController.cls [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        return [SELECT Id, Name, Contact__c, Description__c, Difficulty__c, Due_Date__c, 
+                       Points__c, Record_Type_Name__c, Status__c 
+                FROM Task__c WHERE Status__c != 'Completed'];
+    }
+
+    @AuraEnabled
+    public static List<Task__c> deleteHabitTask(Task__c task) {    
+        List<Task__c> habitTaskList = new List<Task__c>();
```
Comment (rjulianialtimetrik) [inline src/classes/TaskController.cls]: This method is fine but should be more generic. just "deleteTask". Also, it should not do a query or return a list.

File: src/classes/TaskController.cls [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            throw new System.NoAccessException(); 
+            return null;              
+      } 
+        delete task;
+        return habitTaskList = getHabitTaskList();   
+    }
+
+    @AuraEnabled
+    public static List<Task__c> updateStatus(Task__c task) {
+       List<Task__c> taskList = new List<Task__c>();
```
Comment (rjulianialtimetrik) [inline src/classes/TaskController.cls]: this method should not perform a query. you should handle the visual update in the front-end.

File: src/classes/TaskController.cls [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        
+        update task;
+        return taskList = getDailyOrToDoTaskList(task.Record_Type_Name__c);
+    }
+
+    //Get the "Habit" tasks
+    private static List<Task__c> getHabitTaskList() {
+        
+        String[] fieldsToCheck = new String[] {
```
Comment (rjulianialtimetrik) [inline src/classes/TaskController.cls]: Why did you define a new method that queries only on habits, but did another one that queries on both To-Do and Daily? 

You should be able to use the latter to query habit.

File: src/classes/UpdateDailyTasks.cls [hunk 14]
```diff
@@ -0,0 +1,35 @@
+public class UpdateDailyTasks {
```
Comment (rjulianialtimetrik) [inline src/classes/UpdateDailyTasks.cls]: What is this whole class being used for?

**Comment by ealbuerne** [comment]
@rjulianialtimetrik Please check the changes based on comments you made. Thank you.

**Pull request merged**